rust/src/main.rs

```diff
@@ -76,32 +69,130 @@ fn setup_logging() -> Result<(), fern::InitError> {
         }).level(level_filter)
         .chain(io::stderr());
 
-    let path_result = path_for_log_file(XI_LOG_FILE);
-    // If the logging_file_path returned successfully, add the logfile capability to fern
-    if let Ok(logging_file_path) = &path_result {
+    if let Ok(logging_file_path) = &logging_path_result {
+        // Ensure the logging directory is created
+        let parent_path = logging_file_path.parent().ok_or(io::Error::new(
+            io::ErrorKind::InvalidInput,
+            format!(
+                "Unable to get the parent of the following Path: {}",
+                logging_file_path.display(),
+            ),
+        ))?;
+        fs::create_dir_all(parent_path)?;
```

Comment:
This recursively creates *all* missing directories so erroring if the parent is missing just above this seems confusing. I guess it would make sense on windows where you might be logging to an external drive that goes missing, but in the more common case `path/with/nonexisting/directories` missing intermediate directories should not be an error IMHO.

I'd like at least some clarification to comments and the error message. In particular a mention that the path here already contains the filename, because otherwise this becomes very confusing (erroring if someone wants to use `/` as their log dir is just weird).

Reply:
### My thought process:

Well I moved the `fs::create_dir_all()` to right next to the `fern::log_file()` bit as it ensures that we've tried to make sure  there's a directory there before trying to write a file inside.
As it was before any old PathBuf could have been passed in and there was no guarantee that we'd tried to make the required directories.

As we're sending in a PathBuf that looks something like the following: `/home/username123/.local/share/xi-core/xi-core.log` and `fs::create_dir_all()` just wants the directory bit.
```
/home/username123/.local/share/xi-core/xi-core.log
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

So 'PathBuf`s `.parent()` will get that bit **but** will return None if the PathBuf is just `/` or a "prefix".
I can't give `fs::create_dir_all()` a `None`.
I turn the option into a `Result` so I can use `?`/`try!`

**Note:** Someone can still have it log out to `/` they just can't try log *directly* to `/`. Fern wants a location for the log *file* to go and as a result the `PathBuf` *must* be something like `/xi-core.log` to log to the `/` directory

I could do something like the following:
```
match logging_file_path.parent() {
    Some(parent_path) => {
            fs::create_dir_all(parent_path)?;
            // Attach it to fern
            fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?);
    }
        None => ()
}
```
Or
```
match logging_file_path.parent() {
    Some(parent_path) => {
            fs::create_dir_all(parent_path)?;
    }
        None => ()
}
// Attach it to fern
fern_dispatch = fern_dispatch.chain(fern::log_file(logging_file_path)?);
```
But then I don't have an Err any more and it will allow people to put a PathBuf of `/` into `fern::logfile` which will *always* fail.

So I could have just done what I did with the conversion to a Result & then using `?`/`try!`.


---

I do agree that I should add some comments to explain what's going on.
If you have any questions or suggestions for the comments I'm also on irc.

Reply:
I swear every hour I swap between flavoring `if let` over `match` when that'd probably be better and vice versa >.<

As the code is now it `setup_logging` will never recive a `PathBuf` of just `/` but if the code is changed `setup_logging` could recive a `PathBuf` of just `/` so that code is to handle that.

I split it into a separate function wrote some decent docs for it but I can't find them now 😭